pkg/resources/management.cattle.io/v3/nodedriver/validator.go

```diff
@@ -0,0 +1,185 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	gvr = schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodedrivers",
+	}
+
+	driverInUse = &admissionv1.AdmissionResponse{
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "This driver is in use by existing nodes and cannot be disabled",
+		},
+		Allowed: false,
+	}
+)
+
+// Validator ValidatingWebhook for management clusters.
+type Validator struct {
+	admitter admitter
+}
+
+type admitter struct {
+	api k8sAPI
+}
+
+// interface that holds the methods necessary to hit the k8s api and fetch
+// whether rke1/2 resources are deleted
+type k8sAPI interface {
+	rke1ResourcesDeleted(ctx context.Context, driver *v3.NodeDriver) (bool, error)
+	rke2ResourcesDeleted(ctx context.Context, driver *v3.NodeDriver) (bool, error)
+}
+
+type k8sAPIImpl struct {
+	rest rest.Interface
+}
+
+func NewValidator(client rest.Interface) admission.ValidatingAdmissionHandler {
+	return &Validator{admitter: admitter{api: &k8sAPIImpl{rest: client}}}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Update, admissionregistrationv1.Delete}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	return []admissionregistrationv1.ValidatingWebhook{*admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.ClusterScope, v.Operations())}
+}
+
+// Admitters returns the admitter objects used to validate machineconfigs.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	old, new, err := objectsv3.NodeDriverOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// hardcoded bypass - if the annotation is present we want to allow the
+	// driver to be disabled.
+	if old.Annotations[overrideAnnotation] == "true" || new.Annotations != nil && new.Annotations[overrideAnnotation] == "true" {
+		return admission.ResponseAllowed(), nil
+	}
+
+	// the check to see if the driver is being disabled is either when we're
+	// running a delete operation OR an update operation where the active flag
+	// toggles from true -> false
+	var disablingDriver bool
+	if request.Operation == admissionv1.Delete && old.Spec.Active {
+		disablingDriver = true
+	} else if !new.Spec.Active && old.Spec.Active {
+		disablingDriver = true
+	}
```

Comment:
Add a bit of future-proofing here if someone were to go and add a new operation for separate logic, we want to make sure this evaluates the same.
```suggestion
	if request.Operation == admissionv1.Delete && old.Spec.Active ||  request.Operation == admissionv1.Update && !new.Spec.Active && old.Spec.Active {
		disablingDriver = true
	}
```